File: packages/astra-core/src/predictive.ts
// predictive.ts — time-based techniques (MOBILE_ROADMAP §3.4), a port of the
// chart-deriving half of backend/predictive.py:
//   • Secondary progressions ("a day for a year")
//   • Solar return (the chart for the Sun's annual return to its natal longitude)
//
// Eclipse timelines are NOT ported here: the backend uses Swiss Ephemeris'
// eclipse-search (sol_eclipse_when_glob / lun_eclipse_when) and its total/
// annular/partial classification, which astronomy-engine reproduces by a
// different algorithm — matching dates + nature cross-engine is the "hard 20%"
// deferred with the Node/Chiron gap (WASM-Swiss escalation, roadmap §3).
//
// Drift-locked to the backend by parity/predictive.json.

import { calculateChart, aspectsBetween, eclipticLonSpeed, julianDay } from "./ephemeris.js";
import type { Angles, Aspect, ChartRequest, HouseCusp, PlanetData } from "./types.js";

const TROPICAL_YEAR = 365.24219; // mean tropical year, days
const SUN_DEG_PER_DAY = 0.9856473;
const MS_PER_DAY = 86_400_000;
const JD_UNIX_EPOCH = 2440587.5;

const round2 = (x: number) => Math.round(x * 100) / 100;

export interface ProgressedChart {
  age_years: number;
  progressed_iso: string;
  planets: PlanetData[];
  aspects_to_natal: Aspect[];
  meta: Record<string, string>;
}

export interface SolarReturnChart {
  year: number;
  return_iso: string;
  planets: PlanetData[];
  houses: HouseCusp[];
  angles: Angles;
  aspects: Aspect[];
  elements: Record<string, number>;
  modalities: Record<string, number>;
  meta: Record<string, string>;
}

// --------------------------------------------------------------------------- //
// Time helpers (UTC throughout)
// --------------------------------------------------------------------------- //

/** Local birth fields minus tz_offset → UTC epoch milliseconds. */
function natalUtcMillis(req: ChartRequest): number {
  const localMs = Date.UTC(
    req.year, req.month - 1, req.day,
    req.hour ?? 0, req.minute ?? 0, req.second ?? 0
  );
  return localMs - (req.tz_offset ?? 0) * 3600 * 1000;
}

/** ISO-8601 → epoch ms, treating a tz-less string as UTC (backend _parse_iso). */
function parseIsoUtc(s: string): number {
  const hasTz = /[zZ]|[+-]\d{2}:?\d{2}$/.test(s.trim());
  return Date.parse(hasTz ? s : `${s}Z`);
}

/** UTC instant → whole-second ISO with a +00:00 suffix (Python isoformat, µs=0). */
function isoUtcSeconds(d: Date): string {
  return `${d.toISOString().slice(0, 19)}+00:00`;
}

function utcToRequest(
  d: Date, lat: number, lng: number, base: ChartRequest
): ChartRequest {
  return {
    year: d.getUTCFullYear(),
    month: d.getUTCMonth() + 1,
    day: d.getUTCDate(),
    hour: d.getUTCHours(),
    minute: d.getUTCMinutes(),
    second: d.getUTCSeconds(),
    lat,
    lng,
    tz_offset: 0,
    house_system: base.house_system ?? "P",
    zodiac: base.zodiac ?? "tropical",
    ayanamsha: base.ayanamsha ?? 1,
  };
}

/** Signed shortest angular distance target-current in (-180, 180]. */
function signedDelta(target: number, current: number): number {
  return ((target - current + 540) % 360) - 180;
}

// --------------------------------------------------------------------------- //
// Secondary progressions
// --------------------------------------------------------------------------- //

export function progressedChart(natal: ChartRequest, targetIso: string): ProgressedChart {
  const natalMs = natalUtcMillis(natal);
  const targetMs = parseIsoUtc(targetIso);
  const ageYears = (targetMs - natalMs) / (TROPICAL_YEAR * MS_PER_DAY);

  // One day of ephemeris time per year of life.
  const progressedMs = natalMs + ageYears * MS_PER_DAY;
  const progD = new Date(progressedMs);
  const prog = calculateChart(utcToRequest(progD, natal.lat, natal.lng, natal));
  const natalChart = calculateChart(natal);

  return {
    age_years: round2(ageYears),
    progressed_iso: isoUtcSeconds(progD),
    planets: prog.planets,
    aspects_to_natal: aspectsBetween(natalChart.planets, prog.planets),
    meta: {
      method: "secondary_progression",
      ephem_iso: progD.toISOString().slice(0, 10),
    },
  };
}

// --------------------------------------------------------------------------- //
// Solar return
// --------------------------------------------------------------------------- //

/** JD nearest the birthday in `year` when the Sun is back at `natalSunLon`. */
export function solarReturnJd(
  natalSunLon: number, year: number, month: number, day: number
): number {
  let jd = julianDay(year, month, day, 12.0);
  for (let i = 0; i < 10; i++) {
    const sun = eclipticLonSpeed(jd, "Sun");
    if (!sun) break;
    const delta = signedDelta(natalSunLon, sun.lon);
    jd += delta / SUN_DEG_PER_DAY;
    if (Math.abs(delta) < 1e-7) break;
  }
  return jd;
}

function jdToUtcDate(jd: number): Date {
  // Round to the nearest whole second (backend _jd_to_utc rounds, not floors).
  return new Date(Math.round((jd - JD_UNIX_EPOCH) * MS_PER_DAY / 1000) * 1000);
}

export function solarReturn(
  natal: ChartRequest, year: number, lat?: number, lng?: number
): SolarReturnChart {
  const natalChart = calculateChart(natal);
  const natalSun = natalChart.planets.find((p) => p.id === "Sun");
  if (!natalSun) throw new Error("natal chart has no Sun");
  const jd = solarReturnJd(natalSun.longitude, year, natal.month, natal.day);
  const retD = jdToUtcDate(jd);
  const req = utcToRequest(retD, lat ?? natal.lat, lng ?? natal.lng, natal);
  const chart = calculateChart(req);
  return {
    year,
    return_iso: isoUtcSeconds(retD),
    planets: chart.planets,
    houses: chart.houses,
    angles: chart.angles,
    aspects: chart.aspects,
    elements: chart.elements,
    modalities: chart.modalities,
    meta: {
      method: "solar_return",
      relocated: String(lat !== undefined || lng !== undefined),
    },
  };
}

File: backend/predictive.py
"""
predictive.py
=============
Time-based techniques layered on the natal chart:

  * Secondary progressions — "a day for a year": advance the birth moment one day
    per year of life, recompute, and read the progressed planets (and their
    aspects back to the natal chart) as inner symbolic development.
  * Solar return — the chart for the exact moment each year that the transiting
    Sun returns to its natal ecliptic longitude; read as that year's theme.
  * Eclipse timeline — upcoming solar/lunar eclipses and which natal points they
    activate (conjunction / opposition within orb).

All symbolic, never deterministic prediction. Uses Swiss Ephemeris directly for
the Sun-longitude root-find and the eclipse search, and reuses ephemeris.py to
build the derived charts.
"""

from __future__ import annotations

import datetime as dt
from typing import List, Optional

import swisseph as swe
from pydantic import BaseModel, Field

import astrology as A
import ephemeris as E
from models import Angles, Aspect, ChartRequest, HouseCusp, PlanetData

# Reuse whatever ephemeris mode is active (Swiss files when SE_EPHE_PATH is set,
# Moshier otherwise) so results match the rest of the app.
_FLAGS = E._FLG_LON
_FLAGS_ECL = swe.FLG_SWIEPH | swe.FLG_MOSEPH

_TROPICAL_YEAR = 365.24219  # mean tropical year, days
_SUN_DEG_PER_DAY = 0.9856473
_SKIP = {"Descendant", "Imum Coeli", "Part of Fortune", "Lilith", "South Node"}

DISCLAIMER = (
    "Progressions, returns, and eclipses are symbolic timing mirrors for reflection, "
    "not deterministic predictions of fixed events."
)


# --------------------------------------------------------------------------- #
# Time helpers (timezone-aware UTC throughout)
# --------------------------------------------------------------------------- #


def _natal_utc(req: ChartRequest) -> dt.datetime:
    local = dt.datetime(req.year, req.month, req.day, req.hour, req.minute, req.second)
    return (local - dt.timedelta(hours=req.tz_offset)).replace(tzinfo=dt.timezone.utc)


def _parse_iso(s: str) -> dt.datetime:
    d = dt.datetime.fromisoformat(s.replace("Z", "+00:00"))
    return d if d.tzinfo else d.replace(tzinfo=dt.timezone.utc)


def _utc_to_request(d: dt.datetime, lat: float, lng: float, *,
                    house_system: str = "P", zodiac: str = "tropical",
                    ayanamsha: int = 1) -> ChartRequest:
    return ChartRequest(
        year=d.year, month=d.month, day=d.day, hour=d.hour, minute=d.minute,
        second=d.second, lat=lat, lng=lng, tz_offset=0.0,
        house_system=house_system, zodiac=zodiac, ayanamsha=ayanamsha,
    )


def _jd(d: dt.datetime) -> float:
    ut = d.hour + d.minute / 60.0 + d.second / 3600.0
    return swe.julday(d.year, d.month, d.day, ut, swe.GREG_CAL)


def _jd_to_utc(jd: float) -> dt.datetime:
    y, m, day, ut = swe.revjul(jd, swe.GREG_CAL)
    # timedelta normalizes the carry when rounding reaches :60 (a plain
    # min(s, 59) clamp silently loses up to a second).
    base = dt.datetime(y, m, day, tzinfo=dt.timezone.utc)
    return base + dt.timedelta(seconds=round(ut * 3600.0))


def _signed_delta(target: float, current: float) -> float:
    """Signed shortest angular distance target-current in (-180, 180]."""
    return ((target - current + 540.0) % 360.0) - 180.0


# --------------------------------------------------------------------------- #
# Models
# --------------------------------------------------------------------------- #


class ProgressedRequest(BaseModel):
    natal: ChartRequest
    target_iso: str = Field(..., description="Date to progress to, ISO-8601")


class ProgressedChart(BaseModel):
    age_years: float
    progressed_iso: str
    planets: List[PlanetData]
    aspects_to_natal: List[Aspect]
    disclaimer: str = DISCLAIMER
    meta: dict = Field(default_factory=dict)


class SolarReturnRequest(BaseModel):
    natal: ChartRequest
    year: int = Field(..., ge=-3000, le=3000)
    # Optional relocation; defaults to the birthplace.
    lat: Optional[float] = None
    lng: Optional[float] = None


class SolarReturnChart(BaseModel):
    year: int
    return_iso: str
    planets: List[PlanetData]
    houses: List[HouseCusp]
    angles: Angles
    aspects: List[Aspect]
    elements: dict
    modalities: dict
    disclaimer: str = DISCLAIMER
    meta: dict = Field(default_factory=dict)


class EclipseRequest(BaseModel):
    natal: ChartRequest
    start_iso: Optional[str] = None     # default: today
    count: int = Field(8, ge=1, le=40)  # number of eclipses to return


class EclipseContact(BaseModel):
    natal_body: str
    aspect: str          # "conjunction" | "opposition"
    orb: float


class EclipseEvent(BaseModel):
    date: str
    kind: str            # "solar" | "lunar"
    nature: str          # "total" | "annular" | "partial" | "penumbral" | ...
    longitude: float
    sign: str
    degree: int
    activations: List[EclipseContact] = Field(default_factory=list)


class EclipseTimelineResponse(BaseModel):
    start: str
    eclipses: List[EclipseEvent]
    disclaimer: str = DISCLAIMER


# --------------------------------------------------------------------------- #
# Secondary progressions
# --------------------------------------------------------------------------- #


def progressed_chart(natal: ChartRequest, target_iso: str) -> ProgressedChart:
    """Day-for-a-year secondary progression to `target_iso`."""
    natal_utc = _natal_utc(natal)
    target = _parse_iso(target_iso)
    age_years = (target - natal_utc).total_seconds() / (_TROPICAL_YEAR * 86400.0)

    # One day of ephemeris time per year of life.
    progressed_utc = natal_utc + dt.timedelta(days=age_years)
    prog_req = _utc_to_request(
        progressed_utc, natal.lat, natal.lng,
        house_system=natal.house_system, zodiac=natal.zodiac, ayanamsha=natal.ayanamsha,
    )
    prog = E.calculate_chart(prog_req)
    natal_chart = E.calculate_chart(natal)

    # Aspects from progressed planets back to the natal chart.
    aspects = E.aspects_between(natal_chart.planets, prog.planets)
    return ProgressedChart(
        age_years=round(age_years, 2),
        progressed_iso=progressed_utc.replace(microsecond=0).isoformat(),
        planets=prog.planets,
        aspects_to_natal=aspects,
        meta={"method": "secondary_progression", "ephem_iso": progressed_utc.date().isoformat()},
    )


# --------------------------------------------------------------------------- #
# Solar return
# --------------------------------------------------------------------------- #


def _sun_longitude(jd: float, flags: int = _FLAGS) -> float:
    return float(swe.calc_ut(jd, swe.SUN, flags)[0][0])


def solar_return_jd(natal_sun_lon: float, year: int, month: int, day: int,
                    flags: int = _FLAGS) -> float:
    """JD nearest the birthday in `year` when the Sun is at `natal_sun_lon`.

    `flags` must put the transiting Sun in the SAME zodiac frame as
    `natal_sun_lon` — comparing a sidereal natal longitude against tropical
    positions offsets the root-find by the ayanamsha (~24 days of Sun motion).
    """
    jd = swe.julday(year, month, day, 12.0, swe.GREG_CAL)
    for _ in range(10):
        delta = _signed_delta(natal_sun_lon, _sun_longitude(jd, flags))
        jd += delta / _SUN_DEG_PER_DAY
        if abs(delta) < 1e-7:
            break
    return jd


def solar_return(natal: ChartRequest, year: int,
                 lat: Optional[float] = None, lng: Optional[float] = None) -> SolarReturnChart:
    natal_chart = E.calculate_chart(natal)
    natal_sun = next(p for p in natal_chart.planets if p.id == "Sun")
    with E.swe_lock:
        # Root-find in the chart's own frame (sidereal natal Sun ⇒ sidereal
        # transiting Sun); the lock keeps the sid-mode set for the whole search.
        sid_flag = E._apply_zodiac(natal)
        jd = solar_return_jd(natal_sun.longitude, year, natal.month, natal.day,
                             _FLAGS | sid_flag)
    ret_utc = _jd_to_utc(jd)
    req = _utc_to_request(
        ret_utc,
        natal.lat if lat is None else lat,
        natal.lng if lng is None else lng,
        house_system=natal.house_system, zodiac=natal.zodiac, ayanamsha=natal.ayanamsha,
    )
    chart = E.calculate_chart(req)
    return SolarReturnChart(
        year=year,
        return_iso=ret_utc.replace(microsecond=0).isoformat(),
        planets=chart.planets, houses=chart.houses, angles=chart.angles,
        aspects=chart.aspects, elements=chart.elements, modalities=chart.modalities,
        meta={"method": "solar_return",
              "relocated": str(lat is not None or lng is not None)},
    )


# --------------------------------------------------------------------------- #
# Eclipse timeline
# --------------------------------------------------------------------------- #

_ECLIPSE_NATURE = [
    ("total", getattr(swe, "SE_ECL_TOTAL", 4)),
    ("annular_total", getattr(swe, "SE_ECL_ANNULAR_TOTAL", 32)),
    ("annular", getattr(swe, "SE_ECL_ANNULAR", 8)),
    ("partial", getattr(swe, "SE_ECL_PARTIAL", 16)),
    ("penumbral", getattr(swe, "SE_ECL_PENUMBRAL", 64)),
]
_ECLIPSE_ORB = 3.0


def _eclipse_nature(retflag: int) -> str:
    for name, bit in _ECLIPSE_NATURE:
        if retflag & bit:
            return name
    return "unknown"


def _eclipse_activations(eclipse_lon: float, natal_planets: List[PlanetData]) -> List[EclipseContact]:
    contacts: List[EclipseContact] = []
    for p in natal_planets:
        if p.id in _SKIP:
            continue
        sep = A.angular_separation(p.longitude, eclipse_lon)
        if sep <= _ECLIPSE_ORB:
            contacts.append(EclipseContact(natal_body=p.id, aspect="conjunction", orb=round(sep, 2)))
        elif abs(sep - 180.0) <= _ECLIPSE_ORB:
            contacts.append(EclipseContact(natal_body=p.id, aspect="opposition", orb=round(abs(sep - 180.0), 2)))
    contacts.sort(key=lambda c: c.orb)
    return contacts


def eclipse_timeline(natal: ChartRequest, start_iso: Optional[str] = None,
                     count: int = 8) -> EclipseTimelineResponse:
    natal_chart = E.calculate_chart(natal)
    start = _parse_iso(start_iso) if start_iso else dt.datetime.now(dt.timezone.utc)
    jd0 = _jd(start)

    found: List[tuple] = []   # (jd, retflag, kind)
    # Solar eclipses
    cur = jd0
    for _ in range(count):
        retflag, tret = swe.sol_eclipse_when_glob(cur, _FLAGS_ECL, 0, False)
        found.append((tret[0], retflag, "solar"))
        cur = tret[0] + 1.0
    # Lunar eclipses
    cur = jd0
    for _ in range(count):
        retflag, tret = swe.lun_eclipse_when(cur, _FLAGS_ECL, 0, False)
        found.append((tret[0], retflag, "lunar"))
        cur = tret[0] + 1.0

    found.sort(key=lambda t: t[0])
    events: List[EclipseEvent] = []
    with E.swe_lock:
        # The eclipse longitude is compared against (and displayed beside) the
        # natal chart's positions — compute it in the chart's zodiac frame.
        sid_flag = E._apply_zodiac(natal)
        lums = {jd_e: float(swe.calc_ut(jd_e, swe.SUN if kind == "solar" else swe.MOON,
                                        _FLAGS | sid_flag)[0][0])
                for jd_e, _retflag, kind in found[:count]}
    for jd_e, retflag, kind in found[:count]:
        lon = lums[jd_e]
        deg, _minute, _sec = A.degree_in_sign(lon)
        events.append(EclipseEvent(
            date=_jd_to_utc(jd_e).date().isoformat(),
            kind=kind, nature=_eclipse_nature(retflag),
            longitude=round(lon, 4), sign=A.sign_for(lon), degree=deg,
            activations=_eclipse_activations(lon, natal_chart.planets),
        ))
    return EclipseTimelineResponse(
        start=start.date().isoformat(), eclipses=events,
    )
